### ISSUE TITLE
Cache BuildKit state volume

### ISSUE THREAD

**Issue by crazy-max** [docker/setup-buildx-action#122]
follow-up https://github.com/docker/build-push-action/issues/482

Since docker/buildx#672 BuildKit state is saved in a named volume. We should be able to cache it and mount it back. This would require the possiblity to name the builder instead of a generated one.

**Comment by basicdays** [comment]
This would be insanely awesome. Would this also cover the situation of not having to set `cache-from` and `cache-to`? My understanding is that image layer caching by default is kept in buildkit?

**Comment by tmc** [comment]
This could be huge for speeding up larger or more complicated builds.

**Comment by crazy-max** [comment]
> This would be insanely awesome. Would this also cover the situation of not having to set `cache-from` and `cache-to`? My understanding is that image layer caching by default is kept in buildkit?

Yes cache exporter would not be required in this case. Note that this solution would only work with GitHub Actions and a defined builder as this is a the local state.

**Comment by Richard87** [comment]
Imagine of much compute resources would be saved if everyone building docker images on github saved 30sec each time 🤯

**Comment by devminded** [comment]
@Richard87 in this article they saved way more than 30 seconds.
https://evilmartians.com/chronicles/build-images-on-github-actions-with-docker-layer-caching

**Comment by Richard87** [comment]
Thanks @devminded ! I remember I already do thos tings, uts jjust when I update a node dependency or PHP dependencies that everything takes forever 🤔

**Comment by Drowze** [comment]
This is very desired feature 😄 

For our ruby applications, our Dockerfile look sorta like:
```Dockerfile
FROM alpine-ruby:latest
RUN mkdir /app
COPY Gemfile Gemfile.lock .ruby-version /app
RUN bundle install

COPY . /app
```

...so any change to our installed dependencies (described on `Gemfile` and `Gemfile.lock`) invalidates the layer that actually install the dependencies (`bundle install`), meaning we fetch and install _all_ dependencies if a single dependency is changed.

I'm sure this is a very common scenario and, if it was possible to cache the dependencies directory in a docker volume and reuse that docker volume to build new images in between triggered pipelines, I'm sure the impact would be huge!

Very much looking forward to this 👍

**Comment by crazy-max** [comment]
@Drowze

> if it was possible to cache the dependencies directory

You can already cache dependencies using the [`RUN --mount=type=cache` mount](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache). See this example: https://github.com/portapps/portapps.github.io/blob/cde4dd6de0a2876f0311ae321f25f454cf417076/dev.Dockerfile#L7-L13

**Comment by Drowze** [comment]
> You can already cache dependencies using the [`RUN --mount=type=cache` mount](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache). See this example: https://github.com/portapps/portapps.github.io/blob/cde4dd6de0a2876f0311ae321f25f454cf417076/dev.Dockerfile#L7-L13

So using `RUN --mount=type=cache` will cache that mount in between builds using `gha` caching type? Even on GitHub managed runners (*not* self-hosted)? 🤔 
I'll give it a try. Thanks for pointing me that @crazy-max 😄

**Comment by crazy-max** [comment]
> So using `RUN --mount=type=cache` will cache that mount in between builds using `gha` caching type? Even on GitHub managed runners (_not_ self-hosted)? 🤔

No it will not exporter cache (cache-to) but it will cache to the buildkit state (volume).

**Comment by wille** [comment]
setup-buildx-action is unusable in larger builds because of the lack for buildkit state caching support

**Comment by ibalajiarun** [comment]
I have this PR out to address this: https://github.com/docker/setup-buildx-action/pull/211

**Comment by crazy-max** [comment]
> I have this PR out to address this: #211

Keeping BuildKit state will not work on public runners, hence https://github.com/docker/setup-buildx-action/pull/138 to try to address this.

**Comment by ibalajiarun** [comment]
Yes, thats is true. Thanks for clarification. #211 will only work with persistent self-hosted runners.

**Comment by crohr** [comment]
@crazy-max hello, what can we do to make this feature happen? would go hand in hand with https://github.com/docker/setup-buildx-action/pull/427.

> Keeping BuildKit state will not work on public runners

A lot of runner providers now offer snapshotting / sticky disks, etc. Having a way to name the builder would be really nice so that we can do some magic around it (e.g. pre-create / re-hydrate the state volume before buildkit starts, stuff like that).
